### PR TITLE
Added logit/reciprocal decomps, fixed var for complex, moved type promotion logic to standardize on primtorch's

### PR DESCRIPTION
**Pull request by Chillee**

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -579,6 +579,30 @@ def native_dropout_backward(grad_output: Tensor, mask: Tensor, scale: float):
...
+
+@register_decomposition(aten.logit)
+def logit(self: Tensor, eps: Optional[float] = None) -> Tensor:
+    self = _promote_to_float(self)
+    if eps is None:
+        eps = -1.0
+    lo = eps
+    hi = 1 - eps
+    self = torch.clamp(self, lo, hi)
+    print(self.dtype)
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:602]: pront

**Review by ezyang — APPROVED** [review]

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -579,6 +579,30 @@ def native_dropout_backward(grad_output: Tensor, mask: Tensor, scale: float):
     return grad_output * (mask.type_as(grad_output) * scale)
 
 
+@register_decomposition(aten.reciprocal)
+def reciprocal(self: Tensor) -> Tensor:
+    return 1 / self
+
+
+# Used for implementing ops that implicitly promote ints to floats. Could be in Primtorch refs?
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:587]: heh, it not just promotes int to float, it also makes sure that intermediate math is done in fp32? Logit core implementation does that.

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (ngimel) [reply torch/_decomp/decompositions.py:602]: So what is it for half inputs? Looks like it should be float, and you'd return float that you didn't expect?

**Comment by Chillee** [comment]
So, I changed the type promotion logic to re-use primtorch's - would appreciate a re-review @ezyang @ngimel.

There's some places where we're currently re-using primtorch's "elementwise_dtypes" promotion rules where they're not ... quite right (such as native_layer_norm, but that's always been true :P).

**Comment by ezyang** [comment]
update lgtm

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -20,38 +22,36 @@ class Reduction(Enum):
     SUM = 2
 
 
-def cast_for_opmath(f):
+# This wraps a decomposition and performs various type promotion logic within it, depending on the strategy provided
```
Comment (mruberry) [inline torch/_decomp/decompositions.py:25]: Nice comment

**Comment by mruberry** [comment]
@Chillee let me know if some changes to existing dtype/type utilities or new ones would be helpful

**Comment by Chillee** [comment]
@mruberry Some thoughts:

1. Perhaps we should move the elementwise type promotion to `utils`. Alternately, we could potentially move towards having the ATen decomps and the "refs" live in the same file? It's a bit awkward to import `_elementwise_dtypes` from `refs`. Perhaps @ezyang is already fixing this?
2. Speaking of `_elementwise_dtypes`, is it just for elementwise ops or do other non-elementwise ops have similar semantics? (mostly trying to understand when I can re-use this). 
3. Some other type promotion semantics I've needed/run into (I'd be glad to write some of these, although happy to hear some input on how these should be implemented).

`embedding_dense_backward` takes in a tensor with integer dtype (among other floating point tensors). However, some of its intermediate computations need do be done in fp32 (at least - it has an `index_put` with `accumulate=True`). In general, perhaps we need some way of specifying that 

losses (like `huber_loss`) appear to essentially cast to the dtype of the `self` argument. For example, `huber_loss(float32, float64) => float32`. 

Norms have kinda tricky type promotion semantics in general, but these might just need to be implemented on a case-by-case basis.

**Comment by mruberry** [comment]
> @mruberry Some thoughts:
> 
> 1. Perhaps we should move the elementwise type promotion to `utils`. Alternately, we could potentially move towards having the ATen decomps and the "refs" live in the same file? It's a bit awkward to import `_elementwise_dtypes` from `refs`. Perhaps @ezyang is already fixing this?

Moving it to utils sounds good. I can take care of that in a PR shortly.

> 2. Speaking of `_elementwise_dtypes`, is it just for elementwise ops or do other non-elementwise ops have similar semantics? (mostly trying to understand when I can re-use this).

It is intended just for elementwise types. Operators that aren't elementwise or reductions (which have their own type promotion logic) typically shouldn't implement type promotion. See details in https://github.com/pytorch/pytorch/issues/56356. That said, there are some operators like cat which do support type promotion and aren't typically considered "elementwise," and the elementwise type promotion operation should return the correct dtype for it, too.

> 3. Some other type promotion semantics I've needed/run into (I'd be glad to write some of these, although happy to hear some input on how these should be implemented).
> 
> `embedding_dense_backward` takes in a tensor with integer dtype (among other floating point tensors). However, some of its intermediate computations need do be done in fp32 (at least - it has an `index_put` with `accumulate=True`). In general, perhaps we need some way of specifying that

There's a INT_TO_FLOAT type promotion kind but if more specific manipulation is needed it's probably best implemented within the operator itself. 
> 
> losses (like `huber_loss`) appear to essentially cast to the dtype of the `self` argument. For example, `huber_loss(float32, float64) => float32`.

That sounds like a bug.

> Norms have kinda tricky type promotion semantics in general, but these might just need to be implemented on a case-by-case basis.

Yeah I wouldn't want to create a new type promotion kind unless there were a few operators that participate in it.

**Comment by ezyang** [comment]
> Perhaps we should move the elementwise type promotion to utils. Alternately, we could potentially move towards having the ATen decomps and the "refs" live in the same file? It's a bit awkward to import _elementwise_dtypes from refs. Perhaps @ezyang is already fixing this?

I'm not, I am just wiring stuff up, haven't come up with a code organization yet.

**Comment by Chillee** [comment]
Only failing test is backwards_compat test, which is failing on master (and not related).

Crossing my fingers.... 

@pytorchbot  merge this please

File: torch/_decomp/decompositions.py [hunk 4]
```diff
@@ -20,38 +22,36 @@ class Reduction(Enum):
...
         def increase_prec(x):
-            if isinstance(x, Tensor) and x.dtype in (torch.float16, torch.bfloat16):
-                nonlocal orig_prec
-                if orig_prec is None:
-                    orig_prec = x.dtype
-                else:
-                    assert orig_prec == x.dtype
-                return x.to(torch.float32)
+            if isinstance(x, Tensor):
+                return x.to(computation_dtype)
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:37]: This would mask possible type errors for ops that don't support inputs of the different type?

Comment (Chillee) [reply torch/_decomp/decompositions.py:37]: yes, for now. I was thinking another wrapper on top of it for situations like that (or for situations where one tensor needs to be say, an integer dtype).

Any thoughts?
